### PR TITLE
Deallocation date functionality

### PR DESCRIPTION
**Pull request by JohnFarrellDev**
## Link to JIRA ticket

[Change the Service API's allocations endpoint so it uses the provided values](https://hackney.atlassian.net/jira/software/projects/SCS/boards/51?selectedIssue=SCS-632)

### *What is the problem we're trying to solve*

When deallocating a worker we want to set a deallocation date, this date must be in the past.

### *What changes have we introduced*

Updated controller to accept a new property for the date time.
Validate date time is in the past.
Update gateway to ensure the requested deallocation date is used.

#### _Checklist_

- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Get into staging.
Test to see if it is working.
Update feature flags and coordinate release with FE.

### CODE REVIEW

File: SocialCareCaseViewerApi.Tests/V1/Controllers/SocialCareCaseViewerApiControllerTests.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 using AutoFixture;
 using FluentAssertions;
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Diagnostics;
```
Comment (LBHTKarki) [inline SocialCareCaseViewerApi.Tests/V1/Controllers/SocialCareCaseViewerApiControllerTests.cs:6]: is this required?

File: SocialCareCaseViewerApi.Tests/V1/Helpers/TestHelpers.cs [hunk 2]
```diff
@@ -128,6 +128,27 @@ public static ResidentHistoricRecordVisit CreateResidentHistoricRecordVisit(long
...
+            var person = CreatePerson();
+            var team = CreateTeam();
+
+            var updateAllocationRequest = new Faker<UpdateAllocationRequest>()
+                .RuleFor(u => u.Id, f => id ?? f.UniqueIndex + 1)
+                .RuleFor(u => u.DeallocationReason, f => deallocationReason ?? f.Random.String2(200))
+                .RuleFor(u => u.CreatedBy, createdBy ?? updatedByWorker.Email)
+                .RuleFor(u => u.DeallocationDate, f => deallocationDate ?? f.Date.Recent());
+
+            return (updateAllocationRequest, worker, updatedByWorker, person, team);
```
Comment (LBHTKarki) [inline SocialCareCaseViewerApi.Tests/V1/Helpers/TestHelpers.cs:149]: this is neat

File: SocialCareCaseViewerApi/V1/Boundary/Requests/UpdateAllocationRequest.cs [hunk 3]
```diff
@@ -1,24 +1,37 @@
...
 
-        [FromBody]
-        [Required]
-        [EmailAddress]
         public string CreatedBy { get; set; }
+
+        public DateTime? DeallocationDate { get; set; }
+    }
+
+    public class UpdateAllocationRequestValidator : AbstractValidator<UpdateAllocationRequest>
```
Comment (LBHTKarki) [inline SocialCareCaseViewerApi/V1/Boundary/Requests/UpdateAllocationRequest.cs:17]: 👍 for implementing this

File: terraform/production/feature_flags.tf [hunk 4]
```diff
@@ -6,3 +6,12 @@ resource "aws_secretsmanager_secret_version" "show_historic_data_feature_flag" {
   secret_id     = aws_secretsmanager_secret.show_historic_data_feature_flag.id
   secret_string = "true"
 }
+
+resource "aws_secretsmanager_secret" "deallocation_date_feature_flag" {
```
Comment (LBHTKarki) [inline terraform/production/feature_flags.tf:10]: Please don't enable or run terraform for production for now. This is due to reasons discussed outside this PR

**Review by LBHTKarki — APPROVED** [review]

File: terraform/production/feature_flags.tf (hunk 4 above)
Comment (chubberlisk) [reply terraform/production/feature_flags.tf]: ^ We'll need to create these manually through the console as like Tuomo said, we can't / don't want to be running Terraform on production.

File: terraform/staging/feature_flags.tf [hunk 5]
```diff
@@ -6,3 +6,12 @@ resource "aws_secretsmanager_secret_version" "show_historic_data_feature_flag" {
...
 }
+
+resource "aws_secretsmanager_secret" "deallocation_date_feature_flag" {
+    name = "social_care_case_viewer_api_deallocation_date"
+}
+
+resource "aws_secretsmanager_secret_version" "social_care_case_viewer_api_deallocation_date" {
+    secret_id     = aws_secretsmanager_secret.deallocation_date_feature_flag.id
+    secret_string = "true"
+}
```
Comment (chubberlisk) [inline terraform/staging/feature_flags.tf]: If we want to use these values as environment variables, we'll need to update the serverless config for `SOCIAL_CARE_DEALLOCATION_DATE` like what we've done for historic data:

https://github.com/LBHackney-IT/social-care-case-viewer-api/blob/427aa8b04dcdd99fa92b4718eb7089dadb482f80/serverless.yml#L38

**Review by chubberlisk — COMMENTED** [review]
Sorry, wasn't able to review in depth but I noticed something about the feature flags.

**Pull request merged**